### PR TITLE
[KOGITO-4691] Process Marshalling refactor

### PR DESCRIPTION
**Pull request by elguardian**
Jira : https://issues.redhat.com/browse/KOGITO-4691

### CODE REVIEW

**Comment by elguardian** [comment]
@cristianonicolai still pending kogito-apps and kogito-examples.

**Comment by elguardian** [comment]
Jenkins retest this please

**Comment by elguardian** [comment]
Jenkins retest this please

**Comment by elguardian** [comment]
Jenkins retest this please

**Comment by elguardian** [comment]
Jenkins retest this please

**Comment by elguardian** [comment]
Jenkins retest this please

File: kogito-codegen-modules/kogito-codegen-processes/src/test/java/org/kie/kogito/codegen/process/persistence/PostgrePersistenceGeneratorTest.java [hunk 1]
```diff
@@ -56,7 +56,7 @@ void test() {
                 protoGenerator);
         Collection<GeneratedFile> generatedFiles = persistenceGenerator.generate();
 
-        assertThat(generatedFiles).hasSize(5);
+        assertThat(generatedFiles).hasSize(16);
```
Comment (danielezonca) [inline kogito-codegen-modules/kogito-codegen-processes/src/test/java/org/kie/kogito/codegen/process/persistence/PostgrePersistenceGeneratorTest.java:59]: Why now it produces `16` files?

**Review by danielezonca — COMMENTED** [review]
In general it looks good to me, just some comments

**Comment by elguardian** [comment]
seems an error somewhere else 
	[error]: Build step org.kie.kogito.quarkus.common.deployment.KogitoAssetsProcessor#generateModel threw an exception: java.lang.NoSuchMethodError: org.drools.modelcompiler.builder.ModelBuilderImpl.<init>(Ljava/util/function/Function;Lorg/drools/compiler/builder/impl/KnowledgeBuilderConfigurationImpl;Lorg/kie/api/builder/ReleaseId;Z)V

imcompatibility libs

**Comment by elguardian** [comment]
Jenkins retest this please

**Comment by elguardian** [comment]
same probelm in rules
java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
[error]: Build step org.kie.kogito.quarkus.common.deployment.KogitoAssetsProcessor#generateModel threw an exception: java.lang.NoSuchMethodError: 'void org.drools.modelcompiler.builder.ModelBuilderImpl.<init>(java.util.function.Function, org.drools.compiler.builder.impl.KnowledgeBuilderConfigurationImpl, org.kie.api.builder.ReleaseId, boolean)'
at org.kie.kogito.codegen.rules.IncrementalRuleCodegen.generate(IncrementalRuleCodegen.java:156)
at org.kie.kogito.codegen.rules.IncrementalRuleCodegen.generate(IncrementalRuleCodegen.java:85)

@danielezonca  any idea ?

**Comment by danielezonca** [comment]
> NoSuchMethodError

It looks very strange to me, that constructor has been changed ~1 month ago and we recently bumped Kogito to 7.53 that has the new change. It seems something is wrong in the dependency tree of that class (probably still including 7.52 artifacts).
Can you please merge from master in both this PR and https://github.com/kiegroup/kogito-apps/pull/782 to see if it solves the issue?
Btw this should never happen because CI perform a merge from master as first so it is very strange 🤔

**Comment by elguardian** [comment]
@danielezonca  let me try again... but rebased yestarday bump into the error. that is the reason I am askin. Let me check whether there is a problem with some old dependencies or not

**Comment by elguardian** [comment]
Jenkins retest this please

**Comment by danielezonca** [comment]
Ok now it failed on persistence marshalling so it seems more related to this PR

https://rhba-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/blue/organizations/jenkins/KIE%2Fkogito%2Fpullrequest%2Fkogito-runtimes/detail/kogito-runtimes/1531/tests

File: kogito-bom/pom.xml [hunk 2]
```diff
@@ -401,15 +401,9 @@
...
-        <artifactId>protobuf-persistence-common</artifactId>
+        <artifactId>protobuf-serialization-protobuf</artifactId>
         <version>${project.version}</version>
       </dependency>
-      <dependency>
-        <groupId>org.kie.kogito</groupId>
-        <artifactId>protobuf-persistence-common</artifactId>
-        <version>${project.version}</version>
-        <classifier>sources</classifier>
-      </dependency>
```
[the remark this answers is not in the document]
Comment (elguardian) [reply kogito-bom/pom.xml:412]: done

File: jbpm/process-serialization-protobuf/src/main/resources/META-INF/kie.conf [hunk 3]
```diff
@@ -1,2 +1,2 @@
-org.drools.core.marshalling.impl.ProcessMarshallerFactoryService=org.jbpm.marshalling.impl.ProcessMarshallerFactoryServiceImpl
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm/process-serialization-protobuf/src/main/resources/META-INF/kie.conf:1]: done

File: addons/persistence/kafka-persistence-addon/src/main/java/org/kie/kogito/persistence/kafka/KafkaProcessInstances.java [hunk 4]
```diff
@@ -92,7 +90,7 @@ protected void setStore(ReadOnlyKeyValueStore<String, byte[]> store) {
         }
     }
 
-    protected void setMarshaller(ProcessInstanceMarshaller marshaller) {
+    protected void setMarshaller(ProcessInstanceMarshallerService marshaller) {
         this.marshaller = marshaller;
     }
```
[the remark this answers is not in the document]
Comment (elguardian) [reply addons/persistence/kafka-persistence-addon/src/main/java/org/kie/kogito/persistence/kafka/KafkaProcessInstances.java:95]: seems so.

File: kogito-codegen-modules/kogito-codegen-processes/src/test/java/org/kie/kogito/codegen/process/persistence/PostgrePersistenceGeneratorTest.java (hunk 1 above)
Comment (elguardian) [reply kogito-codegen-modules/kogito-codegen-processes/src/test/java/org/kie/kogito/codegen/process/persistence/PostgrePersistenceGeneratorTest.java:59]: protostream now are generated as well as protostream base marshaller have been migrated to codegen because it belongs to the user types.

**Comment by elguardian** [comment]
timed out ??
Jenkins retest this please

File: jbpm/process-serialization-protobuf/pom.xml [hunk 5]
```diff
@@ -14,39 +14,56 @@
...
+    </dependency>
+
+
   </dependencies>
+  <build>
+    <extensions>
+        <extension>
+            <groupId>kr.motd.maven</groupId>
+            <artifactId>os-maven-plugin</artifactId>
+            <version>1.6.0</version>
```
Comment (cristianonicolai) [inline jbpm/process-serialization-protobuf/pom.xml:47]: @elguardian can you add this as managed version in the build-parent-pom?

File: jbpm/process-serialization-protobuf/pom.xml [hunk 6]
```diff
@@ -56,13 +73,14 @@
             <artifactId>protobuf-maven-plugin</artifactId>
             <version>0.6.1</version>
             <configuration>
-              <protocArtifact>com.google.protobuf:protoc:3.12.0:exe:</protocArtifact>
-              <protoSourceRoot>${basedir}/src/main/resources/org/jbpm/marshalling</protoSourceRoot>
+              <protocArtifact>com.google.protobuf:protoc:3.15.0:exe:${os.detected.classifier}</protocArtifact>
```
Comment (cristianonicolai) [inline jbpm/process-serialization-protobuf/pom.xml:76]: regarding the version here, would it make sense to reuse/move to a property in kogito-build-parent?

File: jbpm/process-serialization-protobuf/src/main/java/org/kie/kogito/serialization/process/impl/ProtobufProcessInstanceMarshaller.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+import org.kie.kogito.serialization.process.ProcessInstanceMarshaller;
+
+/**
+ * Marshaller class for RuleFlowProcessInstances
+ */
+
+public class ProtobufProcessInstanceMarshaller implements ProcessInstanceMarshaller {
+
+    @Override
+    public void writeProcessInstance(MarshallerWriterContext context, ProcessInstance processInstance) throws IOException {
```
Comment (cristianonicolai) [inline jbpm/process-serialization-protobuf/src/main/java/org/kie/kogito/serialization/process/impl/ProtobufProcessInstanceMarshaller.java:33]: @elguardian in thinking of potentially removing the dep to v7, would it make sense to use _KogitoProcessInstance_ here instead of ProcessInstance. Making just that dependency visible inside the marshaller were strictly needed?

File: jbpm/process-serialization-protobuf/src/main/java/org/kie/kogito/serialization/process/impl/marshallers/ProtobufBooleanMarshallerStrategy.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    @Override
+    public Object unmarshall(Object marshalled) {
+        try {
+            Any data = (Any) marshalled;
+            BoolValue storedValue = data.unpack(BoolValue.class);
+            return storedValue.getValue();
+        } catch (InvalidProtocolBufferException e1) {
+            // do nothing
```
Comment (cristianonicolai) [inline jbpm/process-serialization-protobuf/src/main/java/org/kie/kogito/serialization/process/impl/marshallers/ProtobufBooleanMarshallerStrategy.java:49]: shouldnt we fail in this case?

**Review by cristianonicolai — COMMENTED** [review]
@elguardian very nice, just some minor comments

File: jbpm/process-serialization-protobuf/src/main/java/org/kie/kogito/serialization/process/impl/marshallers/ProtobufBooleanMarshallerStrategy.java (hunk 8 above)
Comment (elguardian) [reply jbpm/process-serialization-protobuf/src/main/java/org/kie/kogito/serialization/process/impl/marshallers/ProtobufBooleanMarshallerStrategy.java]: throw an exception about marshalling problems in every marshaller

File: jbpm/process-serialization-protobuf/src/main/java/org/kie/kogito/serialization/process/impl/ProtobufProcessInstanceMarshaller.java (hunk 7 above)
Comment (elguardian) [reply jbpm/process-serialization-protobuf/src/main/java/org/kie/kogito/serialization/process/impl/ProtobufProcessInstanceMarshaller.java]: I sank those. Now they depends on org.kie.kogito

File: jbpm/process-serialization-protobuf/pom.xml (hunk 6 above)
Comment (elguardian) [reply jbpm/process-serialization-protobuf/pom.xml]: moved

File: jbpm/process-serialization-protobuf/pom.xml (hunk 5 above)
Comment (elguardian) [reply jbpm/process-serialization-protobuf/pom.xml]: moved

**Review by cristianonicolai — APPROVED** [review]

**Review by tiagodolphine — APPROVED** [review]
LGTM, great stuff, thanks @elguardian

**Comment by elguardian** [comment]
rebased... if nothing gets in the middle should be green
